syntax = "proto3";

package envoy.api.v2;

option java_outer_classname = "SrdsProto";
option java_package = "io.envoyproxy.envoy.api.v2";
option java_multiple_files = true;
option java_generic_services = true;

import "envoy/api/v2/discovery.proto";

import "google/api/annotations.proto";

import "validate/validate.proto";
import "gogoproto/gogo.proto";

option (gogoproto.equal_all) = true;

// [#protodoc-title: HTTP scoped routing configuration]
// * Routing :ref:`architecture overview <arch_overview_http_routing>`
//
// .. attention::
//
//   The Scoped RDS API is not yet fully implemented and *should not* be enabled in
//   :ref:`envoy_api_msg_config.filter.network.http_connection_manager.v2.HttpConnectionManager`.
//
// TODO(AndresGuedez): Update :ref:`arch_overview_http_routing` with scoped routing overview and
// configuration details.

// The `ref:resource_names<envoy_api_field_DiscoveryRequest.resource_names>` field in
// DiscoveryRequest specifies a set of route configuration scopes. Each scope points to a
// route_configuration_name which will be used to request a
// :ref:`RouteConfiguration<envoy_api_msg_RouteConfiguration>` via the RDS API.
service ScopedRoutesDiscoveryService {
  rpc StreamScopedRoutes(stream DiscoveryRequest) returns (stream DiscoveryResponse) {
  }

  rpc IncrementalScopedRoutes(stream IncrementalDiscoveryRequest)
      returns (stream IncrementalDiscoveryResponse) {
  }

  rpc FetchScopedRoutes(DiscoveryRequest) returns (DiscoveryResponse) {
    option (google.api.http) = {
      post: "/v2/discovery:scoped-routes"
      body: "*"
    };
  }
}

// This configuration represents a set of routing 'scopes', each containing independent routing
// configuration (see :ref:`routing architecture overview <arch_overview_http_routing>`).
//
// A ScopedRouteConfigurationsSet is associated with an
// :ref:`HttpConnectionManager<envoy_api_msg_config.filter.network.http_connection_manager.v2.HttpConnectionManager>`
// via the
// :ref:`scoped_rds<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.scoped_rds>`
// or
// :ref:`scoped_routes_config<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.scoped_routes_config>`
// fields. When scoped routing is enabled, the Envoy router will first attempt to determine the
// routing scope assigned to a request by building a
// :ref:`Scope.Key<envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` based on request header
// attributes (such as the value of a header field) defined by the
// :ref:`ScopeKeyBuilder<envoy_api_msg_ScopedRouteConfigurationsSet.ScopeKeyBuilder>`
// this message.
//
// The key will be looked up in a table maintained by the router, containing the
// :ref:`Scope.Key<envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` to
// :ref:`RouteConfiguration<envoy_api_msg_RouteConfiguration>` mappings specified by the
// :ref:`scopes<envoy_api_field_ScopedRouteConfigurationsSet.scopes>` field in this message. The
// routing rules specified by the resulting RouteConfiguration will be used to determine the route
// for the request.
//
// For example, with the following YAML configuration:
//
// .. code::
//
//   scope_key_builder:
//     fragments:
//       - header_element:
//         name: X-Route-Selector
//         element_separator: ,
//         element:
//           separator: =
//           key: vip
//   scopes:
//     - route_configuration_name: route-config1
//       key:
//          fragments:
//            - string_key: 172.10.10.20
//      - route_configuration_name: route-config2
//        key:
//          fragments:
//            - string_key: 172.20.20.30
//
// A request from a client such as:
//
// .. code::
//
//     GET / HTTP/1.1
//     Host: foo.com
//     X-Route-Selector: vip=172.10.10.20
//
// Would result in the `route-config1` RouteConfiguration being assigned to the request/stream,
// which Envoy would fetch via RDS.
//
// [#comment:next free field: 4]

// TODO(AndresGuedez):
// Specifies a routing scope, which associates a :ref:`envoy_api_msg_RouteConfiguration`
// (identified by its resource name) to a
// :ref:`Key<envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` which is matched against each
// HTTP request.
message ScopedRouteConfiguration {
  // TODO(AndresGuedez):
  // The name assigned to the routing scope.
  string name = 1 [(validate.rules).string.min_bytes = 1];

  // Specifies a key which is matched against the output of the
  // :ref:`scope_key_builder<envoy_api_field_ScopedRouteConfigurationsSet.scope_key_builder>` to
  // obtain the route_configuration_name/RouteConfiguration associated with the Scope. The
  // matching is dependent on the order of the fragments contained in the Key.
  message Key {
    message Fragment {
      oneof type {
        option (validate.required) = true;

        // A string to match against.
        string string_key = 1;
      }
    }

    // The ordered set of fragments to match against.
    repeated Fragment fragments = 1 [(validate.rules).repeated .min_items = 1];
  }

  // The resource name to use for a :ref:`envoy_api_msg_DiscoveryRequest` to an RDS server to
  // fetch the :ref:`envoy_api_msg_RouteConfiguration` associated with this scope.
  string route_configuration_name = 2 [(validate.rules).string.min_bytes = 1];

  // The key to match against.
  Key key = 3 [(validate.rules).message.required = true];
}
